non-interactive-atomic-nameswaps.md

```diff
@@ -0,0 +1,191 @@
+data is made available.
+
+# Specification
+
+## Data Structures
+
+The data structures and concepts defined below are used throughout the rest of this document.
+
+### Locking Script
+
+A locking script is defined as a tuple of the bech32 version and data. An encoded locking
+script is an address and is specific to a particular network because of the bech32 human
+readable part. The hrp for Handshake mainnet is `hs`.
+
+### Covenant
+
+Handshake introduces covenants to the Bitcoin UTXO model by adding an additional covenant field
+to each output. The covenant field includes a type and covenant data. A UTXO with a particular
+covenant type can only be spent to another covenant type when a spend path between those two types
+is defined by the consensus rules. A state machine of valid spend paths is used to enable
+the on chain auction system for names. The covenant data is used to persist information
+between stages of the auction and includes a nonce to prevent signatures from being reused
+between auctions for the same name.
+
+A UTXO with the covenant type `NONE` can be thought of as being just like a Bitcoin UTXO.
+Covenants involved in the auction process must be linked, meaning that the input index that references a covenant must be at the same output index that updates the covenant.
+
+The covenant is represented as a varint prefixed bytearray of varint prefixed
+covenant items. The number of items and types of the items depend on the covenant type.
+
+### Witness Program
+
+The Witness Program as defined in [BIP 141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#Witness_program) with some slight modifications. Only native witness programs are supported, blake2b160 is used to hash public keys for p2wpkh and sha-3256 is used to hash scripts for p2wsh.
+
+### Swap Script
+
+Handshake builds on Bitcoin Script and introduces a new Opcode called `OP_TYPE` that pushes the type
+of the covenant that is being spent to on to the stack. This can be used to construct a script that
+has different execution paths depending on which covenant spend path the UTXO is being spent to.
+For the counterparty to be able to trust that the name holder will not back out of the swap,
+the ability to cancel the swap needs to be prevented. At the protocol level, this means that
+`UPDATE`, `REVOKE` and `RENEW` must not be able to be spent to. A naive implementation of the script looks like:
+
+```
+OP_TYPE
+0x07 // UPDATE
+OP_EQUAL
+OP_IF
+OP_RETURN
+OP_ENDIF
+OP_TYPE
+0x0b // REVOKE
+OP_EQUAL
+OP_IF
+OP_RETURN
+OP_ENDIF
+OP_TYPE
+0x08 // RENEW
+OP_EQUAL
+OP_IF
+OP_RETURN
+OP_ENDIF
+OP_TYPE
+0x09 // TRANSFER
+OP_EQUAL
+OP_IF
+<public key>
+OP_CHECKSIGVERIFY
+OP_ENDIF
+```
+
+This script could be improved upon and improvements ought to be considered for future versions of this HIP. Improvements include using `OP_CSV` to timeout the swap, enabling the seller to lower
```

Comment:
Do we really need this?

After the name is transferred to the script (with no CSV) and finalised, the seller can create the partially signed transaction described here multiple times, with different values.

The seller should keep this TXs in secret, and only reveal the most expensive one. When the seller decides to go for a discount, he/she can release the cheaper TX.

If there's no buyer at all, the seller can do a regular transfer to his/her own address. The only benefit here would be to restore the ability to revoke the name.